addons/hr_holidays/models/hr_leave.py

```diff
@@ -754,25 +756,35 @@ def _check_approval_update(self, state):
 
             if (state == 'validate1' and val_type == 'both') or (state == 'validate' and val_type == 'manager'):
                 manager = holiday.employee_id.parent_id or holiday.employee_id.department_id.manager_id
-                if (manager and manager != current_employee) and not self.env.user.has_group('hr_holidays.group_hr_holidays_manager'):
+                if not (holiday.employee_id.leave_manager_id and holiday.employee_id.leave_manager_id == self.env.user and is_team_leader) and (manager and manager != current_employee) and not is_manager:
```

Comment:
I haven't tested but it looks like this condition will not work as intended.
By that I mean that you have to be both "leave manager" and be the employee's leave manager.

So if you are manager but not the leave manager of this employee you will not be able to approve it.

I suggest to rewrite the whole method to take in account all the cases (for example if the current user has multiple employees, weird case but we will cover it).


Here is an example, that I haven't run once, but it should help to understand what I mean

```python
def _check_approval_update(self, state):
    """ Check if target state is achievable. """
    # retrieve user's access rights
    is_team_leader = self.env.user.has_group('hr_holidays.group_hr_holidays_team_leader')
    is_officer = self.env.user.has_group('hr_holidays.group_hr_holidays_user')
    is_manager = self.env.user.has_group('hr_holidays.group_hr_holidays_manager')

    # you need to be at least team leader to approve leaves
    if not is_team_leader:
        raise UserError(_('Only a Team Leader, Leave Officer or Manager can approve or refuse leave requests.'))

    # retrieve current user linked employees
    user_employee_ids = self.env['hr.employee'].search([('user_id', '=', self.env.uid)])
    # loop through all the holidays to confirm
    for holiday in self:
        val_type = holiday.holiday_status_id.validation_type
        if state == 'confirm':
            continue

        if state == 'draft':
            if not is_manager and holiday.employee_id not in user_employee_ids:
                raise UserError(_('Only a Leave Manager can reset other people leaves.'))
            continue

        # check the record rules
        holiday.check_access_rule('write')
        # only managers can approve their own leaves
        if holiday.employee_id in user_employee_ids and not is_manager:
            raise UserError(_('Only a Leave Manager can approve its own requests.'))        

        if (state, val_type) in [('validate1', 'both'), ('validate', 'manager')]:
            is_employee_manager_id = (holiday.employee_id.parent_id or holiday.employee_id.department_id.manager_id) in user_employee_ids
            is_employee_leave_manager_id = holiday.employee_id.leave_manager_id == self.env.user
            # if the user has only team leader rights but this is also checked by a record rule
            if not (is_employee_manager_id or is_employee_leave_manager_id) and not is_officer:
                raise UserError(_('You must be either %s\'s manager or Leave officer to approve this leave') % (holiday.employee_id.name))

        if state == 'validate' and val_type == 'both':
            if not is_manager:
                raise UserError(_('Only an Leave Manager can apply the second approval on leave requests.'))
```

And by the way, some cases here do not need to be checked with python code, record rules can do the job since we call `holiday.check_access_rule('write')`

Reply:
I'll have a look at these in more details